src/main/java/org/opensearch/flowframework/workflow/WorkflowProcessSorter.java

```diff
@@ -122,6 +135,59 @@ public List<ProcessNode> sortProcessNodes(Workflow workflow, String workflowId)
         return nodes;
     }
 
+    /**
+     * Validates a sorted workflow, determines if each process node's required plugins are currently installed
+     * @param processNodes A list of process nodes
+     * @throws Exception on validation failure
+     */
+    public void validatePluginsInstalled(List<ProcessNode> processNodes) throws Exception {
+
+        WorkflowValidator validator = WorkflowValidator.parse("mappings/workflow-steps.json");
+
+        // Retrieve node information to ascertain installed plugins
+        NodesInfoRequest nodesInfoRequest = new NodesInfoRequest();
+        nodesInfoRequest.clear().addMetric(NodesInfoRequest.Metric.PLUGINS.metricName());
+        CompletableFuture<List<String>> installedPluginsFuture = new CompletableFuture<>();
+        client.admin().cluster().nodesInfo(nodesInfoRequest, ActionListener.wrap(response -> {
+            List<String> installedPlugins = new ArrayList<>();
+
+            // Retrieve installed plugin names from the local node
+            String localNodeId = clusterService.state().getNodes().getLocalNodeId();
+            NodeInfo info = response.getNodesMap().get(localNodeId);
+            PluginsAndModules plugins = info.getInfo(PluginsAndModules.class);
+            for (PluginInfo pluginInfo : plugins.getPluginInfos()) {
+                installedPlugins.add(pluginInfo.getName());
+            }
+
+            installedPluginsFuture.complete(installedPlugins);
+
+        }, exception -> {
+            logger.error("Failed to retrieve installed plugins");
+            installedPluginsFuture.completeExceptionally(exception);
+        }));
+
+        // Block execution until installed plugin list is returned
+        List<String> installedPlugins = installedPluginsFuture.get();
```

Comment:
1. This blocks.  Use `join()` instead to wait unblocking, and call `get()` after the `join()` passes.

2. If you've completed exceptionally (line 166), this (or `join()`) will throw an exception here when you try to `get()`.  Probably want to catch that and throw a FlowFrameworkException.

Reply:
Actually I'm wrong.  join blocks too.  There's `thenAccept()` to do nonblocking.... but for now let's leave this as-is.